llarp/nodedb.cpp

```diff
@@ -109,7 +109,176 @@ namespace llarp
+      throw std::runtime_error{"Called rotate_rc_source with no connections, does not make sense!"};
+
+    // We should not be in this function if client_known_routers isn't populated
+    if (client_known_routers.size() <= 1)
+      throw std::runtime_error{"Cannot rotate RC source without RC source(s) to rotate to!"};
+
+    RemoteRC new_source{};
+    router.link_manager().get_random_connected(new_source);
+    if (conn_count == 1)
+    {
+      // if we only have one connection, it must be current rc fetch source
+      assert(new_source.router_id() == rc_fetch_source);
+
+      if (pinned_edges.size() == 1)
+      {
+        // only one pinned edge set, use it even though it gave unsatisfactory RCs
+        assert(rc_fetch_source == *(pinned_edges.begin()));
+        log::warning(
+            logcat,
+            "Single pinned edge {} gave bad RC response; still using it despite this.",
+            rc_fetch_source);
+        return;
+      }
+
+      // only one connection, choose a new relay to connect to for rc fetching
+
+      RouterID r = rc_fetch_source;
+      while (r == rc_fetch_source)
+      {
+        std::sample(client_known_routers.begin(), client_known_routers.end(), &r, 1, csrng);
+      }
+      rc_fetch_source = std::move(r);
+      return;
+    }
+
+    // choose one of our other existing connections to use as the RC fetch source
+    while (new_source.router_id() == rc_fetch_source)
+    {
+      router.link_manager().get_random_connected(new_source);
+    }
+    rc_fetch_source = new_source.router_id();
+  }
+
+  // TODO: trust model
+  void
+  NodeDB::ingest_rcs(RouterID source, std::vector<RemoteRC> rcs, rc_time timestamp)
+  {
+    (void)source;
+
+    // TODO: if we don't currently have a "trusted" relay we've been fetching from,
+    // this will be a full list of RCs.  We need to first check if it aligns closely
+    // with our trusted RouterID list, then replace our RCs with the incoming set.
+
+    for (auto& rc : rcs)
+      put_rc_if_newer(std::move(rc), timestamp);
+
+    // TODO: if we have a "trusted" relay we've been fetching from, this will be
+    // an incremental update to the RC list, so *after* insertion we check if the
+    // RCs' RouterIDs closely match our trusted RouterID list.
+
+    last_rc_update_relay_timestamp = timestamp;
+  }
+
+  // TODO: trust model
+  void
+  NodeDB::ingest_router_ids(RouterID source, std::vector<RouterID> ids)
+  {
+    router_id_fetch_responses[source] = std::move(ids);
+
+    router_id_response_count++;
+    if (router_id_response_count == router_id_fetch_sources.size())
+    {
+      // TODO: reconcile all the responses, for now just insert all
+      for (const auto& [rid, responses] : router_id_fetch_responses)
+      {
+        // TODO: empty == failure, handle that case
+        for (const auto& response : responses)
+        {
+          client_known_routers.insert(std::move(response));
+        }
+      }
+      router_id_fetch_in_progress = false;
```

Comment:
Since this boolean is set to true inside `::fetch_router_ids()` and then set to false inside `::ingest_router_ids()` inside a response lambda for `send_control_message`, should we make it atomic?